### ISSUE TITLE
Support mixing constructor and setter binding for `@ConfigurationProperties`

### ISSUE THREAD

**Issue by quaff** [spring-projects/spring-boot#42276]
It's common to divide configuration properties into required and optional, the required properties are defined as `final` and should be bound via constructor, and the optional properties should be bound via setter method but currently ignored.
For example:
```java
@ConfigurationProperties(prefix = "my")
public class MyConfigurationProperties {

	private final String requiredParam1;
	private final String requiredParam2;

	private String optionalParam1;
	private String optionalParam2;

	public MyConfigurationProperties(String requiredParam1, String requiredParam2) {
		this.requiredParam1 = requiredParam1;
		this.requiredParam2 = requiredParam2;
	}

	public String getOptionalParam1() {
		return optionalParam1;
	}

	public void setOptionalParam1(String optionalParam1) {
		this.optionalParam1 = optionalParam1;
	}

	public String getOptionalParam2() {
		return optionalParam2;
	}

	public void setOptionalParam2(String optionalParam2) {
		this.optionalParam2 = optionalParam2;
	}

	public String getRequiredParam1() {
		return requiredParam1;
	}

	public String getRequiredParam2() {
		return requiredParam2;
	}

}
```
and 
```yaml
my:
  required-param1: required-param1
  required-param2: required-param2
  optional-param2: optional-param2
```

`required-param1` and `required-param2` are bound as expected, we should bind `optional-param2` also.

**Comment by quaff** [comment]
> or requiring two passes

I think this approach is better.

> Optional parameters can be constructor bound, and personally I think that's better as it allows the class to remain immutable. Why do you want to mix the two?

That would make the constructor bloat, normally configuration properties are consist of a few required and more optional properties. 
I still suggest to implement it if the cost is not expensive to not surprise people who thought it will works, I can live with that, but we should document that it's not supported, maybe already done?

**Comment by bclozel** [comment]
I don't think we should pursue this, it's going to add complexity to our codebase, probably introduce regressions and degrade performance. I think the two options we're offering are quite nice and I don't see why we should document all other possible options that we don't support.

**Comment by philwebb** [comment]
I'm afraid I don't think we should pursue the either. IMO one of the best benefits of constructor binding is that the resulting object is immutable. I agree that `@DefaultValue` can be a little clunky, but I still think it's better than a half mutable class.

I'm also of the opinion that we should think very carefully before we add more complexity to the binder. Especially to the existing code paths. We've already had a lot of back-compatibilty issues and I'm not keen to introduce more. It's also worth noting that constructor bloat is offset quite significantly when a java `record` is used as you then no longer need to declare all the getters.

Thanks anyway for the suggestion.

**Comment by quaff** [comment]
For people who are interesting in this, I've created prototype, It's a tiny changes, see https://github.com/quaff/spring-boot/tree/patch-71.

**Comment by quaff** [comment]
For performance concern, flag `spring.config.binding-value-object-as-java-bean` introduced and default to `false`, I appreciate that some team member would reconsider https://github.com/spring-projects/spring-boot/compare/main...quaff:patch-71?expand=1

**Comment by philwebb** [comment]
Thanks for the updates, but the performance concern was just one aspect of why I don't think we should make this change. My main issue is still the conceptual complexity this introduces, along with the fact that it prevents immutable configuration property classes.

Flagging to see if anyone on the team disagrees.
